File: Realize/2Week/PracticeStack.py
class MyStack:
    def __init__(self, size):
        self.arr = []
        self.capacity = size
    
    def is_empty(self):
        return len(self.arr) == 0
    
    def is_full(self):
        return len(self.arr) == self.capacity
    
    def push(self, data):
        if self.is_full():
            print("Stack is full!")
            return
        self.arr.append(data)
    
    def pop(self):
        if self.is_empty():
            print("Stack is empty!")
            return None
        return self.arr.pop()
    
    def peek(self):
        if self.is_empty():
            print("Stack is empty!")
            return None
        return self.arr[-1]
    
    def print_stack(self):
        print(" ".join(map(str, self.arr)))

# Test code
if __name__ == "__main__":
    my_stack = MyStack(5)
    print(my_stack.is_empty())  # True
    my_stack.push(1)
    my_stack.push(2)
    my_stack.push(3)
    my_stack.push(4)
    my_stack.push(5)
    my_stack.push(6)  # Stack is full!
    my_stack.print_stack()  # 1 2 3 4 5

    print(my_stack.peek())  # 5
    my_stack.print_stack()  # 1 2 3 4 5

    print(my_stack.pop())  # 5
    print(my_stack.pop())  # 4
    my_stack.print_stack()  # 1 2 3

    print(my_stack.pop())  # 3
    print(my_stack.pop())  # 2
    print(my_stack.pop())  # 1
    my_stack.print_stack()  # (empty output)

File: Realize/2Week/PracticeStack.cpp
#include <iostream>
using namespace std;

class MyStack {
private:
    int* arr;
    int top;
    int capacity;

public:
    MyStack(int size) {
        arr = new int[size];
        capacity = size;
        top = -1;
    }

    ~MyStack() {
        delete[] arr;
    }

    bool isEmpty() {
        return top == -1;
    }

    bool isFull() {
        return top == capacity - 1;
    }

    void push(int data) {
        if (isFull()) {
            cout << "Stack is full!" << endl;
            return;
        }
        arr[++top] = data;
    }

    int pop() {
        if (isEmpty()) {
            cout << "Stack is empty!" << endl;
            return -1;
        }
        return arr[top--];
    }

    int peek() {
        if (isEmpty()) {
            cout << "Stack is empty!" << endl;
            return -1;
        }
        return arr[top];
    }

    void printStack() {
        for (int i = 0; i <= top; i++) {
            cout << arr[i] << " ";
        }
        cout << endl;
    }
};

int main() {
    // Test code
    MyStack myStack(5);
    cout << myStack.isEmpty() << endl; //1 (true라서)
    myStack.push(1);
    myStack.push(2);
    myStack.push(3);
    myStack.push(4);
    myStack.push(5);
    myStack.push(6); // Stack is full!
    myStack.printStack(); // 1 2 3 4 5

    cout << myStack.peek() << endl; // 5
    myStack.printStack(); // 1 2 3 4 5

    cout << myStack.pop() << endl;  // 5
    cout << myStack.pop() << endl;  // 4
    myStack.printStack(); // 1 2 3

    cout << myStack.pop() << endl;  // 3
    cout << myStack.pop() << endl;  // 2
    cout << myStack.pop() << endl;  // 1
    myStack.printStack(); // (empty output)

    return 0;
}
